src/comps/qb.cc

```diff
@@ -0,0 +1,194 @@
+#define VERBOSE
+#define USE_QR
+
+namespace RandLAPACK::comps::qb {
+
+template <typename T>
+void QB<T>::QB2(
+    int64_t m,
+    int64_t n,
+    std::vector<T>& A,
+    int64_t& k,
+    int64_t block_sz,
+    T tol,
+    std::vector<T>& Q,
+    std::vector<T>& B
+){
+    using namespace blas;
+    using namespace lapack;
+
+    int64_t curr_sz = 0;
+    int64_t next_sz = 0;
+
+    T* A_dat = A.data();
+    // pre-compute nrom
+    T norm_A = lange(Norm::Fro, m, n, A_dat, m);
+    // Immediate termination criteria
+    if(norm_A == 0.0)
+    {
+        // Zero matrix termination
+        k = curr_sz;
+        this->termination = 1;
+        return;
+    }
+    
+    // If the space allocated for col in Q and row in B is insufficient for any iterations ...
+    if(std::max( Q.size() / m, B.size() / n) < k)
+    {
+        // ... allocate more!
+        this->curr_lim = std::min(4 * block_sz, k);
+        // No need for data movement in this case
+        RandLAPACK::comps::util::resize<T>(m * this->curr_lim, Q);
+        RandLAPACK::comps::util::resize<T>(this->curr_lim * n, B);
+    }
+    else
+    {
+        this->curr_lim = k;
+    }
+    
+    // Copy the initial data to avoid unwanted modification TODO #1
+    std::vector<T> A_cpy (m * n, 0.0);
+    T* A_cpy_dat = A_cpy.data();
+    lacpy(MatrixType::General, m, n, A_dat, m, A_cpy_dat, m);
+
+    T norm_B = 0.0;
+    T prev_err = 0.0;
+    T approx_err = 0.0;
+
+    T* Q_gram_dat;
+    T* Q_i_gram_dat;
+
+    if(this->orth_check)
+    {
+        Q_gram_dat = RandLAPACK::comps::util::resize(this->curr_lim * this->curr_lim, this->Q_gram);
+        Q_i_gram_dat = RandLAPACK::comps::util::resize(block_sz * block_sz, this->Q_i_gram);
+    }
+
+    T* QtQi_dat = RandLAPACK::comps::util::resize(this->curr_lim * block_sz, this->QtQi);
+    T* Q_i_dat = RandLAPACK::comps::util::resize(m * block_sz, this->Q_i);
+    T* B_i_dat = RandLAPACK::comps::util::resize(block_sz * n, this->B_i);
+
+    T* Q_dat = Q.data();
+    T* B_dat = B.data();
+    
+    while(k > curr_sz)
+    {
+        // Dynamically changing block size
+        block_sz = std::min(block_sz, k - curr_sz);
+        next_sz = curr_sz + block_sz;
+
+        // Make sure we have enough space for everything
+        if(next_sz > this->curr_lim)
+        {
+            this->curr_lim = std::min(2 * this->curr_lim, k);
+            Q_dat = RandLAPACK::comps::util::resize(this->curr_lim * m, Q);
+            B_dat = RandLAPACK::comps::util::row_resize(curr_sz, n, B, this->curr_lim);
+            QtQi_dat = RandLAPACK::comps::util::resize(this->curr_lim * block_sz, QtQi);
+            Q_gram_dat = RandLAPACK::comps::util::resize(this->curr_lim * this->curr_lim, Q_gram);
+        }
+
+        // Calling RangeFinder
+        this->RF_Obj.call(m, n, A_cpy, block_sz, this->Q_i);
+
+        if(this->orth_check)
+        {
+            if (RandLAPACK::comps::util::orthogonality_check(m, block_sz, block_sz, Q_i, Q_i_gram, this->verbosity))
```

Comment:
Same comment as before: can we make this function call more less verbose by an appropriate "import" statement?